### PR TITLE
Issue 3994 : List all Segments in Tier 2

### PR DESCRIPTION
**Pull request by naveenkothamasu**
**Changelog description**
- Added `list-segments` API to retrieve the segments from Tier 2.

*Purpose of the change*
Fixes #[3994](https://github.com/pravega/pravega/issues/3994).

**What the code does**
- Basic implementation of Iterator over the header files in the storage returning `SegmentProperties` per file.
- The iterator logic is tailored to the specific underlying storage depending on how the listing of entries works in that implementation.
-  For retrieving `StreamSegmentInformation` for a particular header file, `RollingSegmentHandle` is used.

**What this doesn't do**
As of now, the API support is only verified for the FileSystemStorage type. For other storage types such as s3 and HDFS, it will be considered later.

**How to verify it**
Unit tests added to verify feature.

### CODE REVIEW

File: bindings/src/main/java/io/pravega/storage/hdfs/HDFSStorage.java [hunk 1]
```diff
@@ -487,6 +491,82 @@ public SegmentHandle create(String streamSegmentName) throws StreamSegmentExcept
...
+                        }
+                        try {
+                            Long.parseLong(fileName.substring(index + 1));
+                        } catch (NumberFormatException nfe) {
+                            return false;
+                        }
+                        return true;
+                    });
+        } catch (FileNotFoundException e) {
+            // TODO log
```
Comment (sachin-j-joshi) [inline bindings/src/main/java/io/pravega/storage/hdfs/HDFSStorage.java]: Please log exception.

File: bindings/src/main/java/io/pravega/storage/hdfs/HDFSStorage.java [hunk 2]
```diff
@@ -487,6 +491,82 @@ public SegmentHandle create(String streamSegmentName) throws StreamSegmentExcept
...
+        } catch (IOException e) {
+            // TODO log
```
Comment (sachin-j-joshi) [inline bindings/src/main/java/io/pravega/storage/hdfs/HDFSStorage.java]: please log exception.

File: segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/StorageTestBase.java [hunk 3]
```diff
@@ -86,6 +86,87 @@ public void testCreate() {
...
+            SegmentHandle handle = s.create(segmentName);
+            if(rand.nextInt(2) == 1) {
+                s.delete(handle);
+                deletedSegments.add(segmentName);
+            }
+        }
+
+        Iterator<SegmentProperties> it = s.listSegments();
+        expectedCount -= deletedSegments.size();
+        Assert.assertEquals(expectedCount, Iterators.size(it));
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/StorageTestBase.java]: Should we check contents of the list returned to make sure deleted files are not present in that list ? (and all that were expected are present)

File: shared/protocol/src/main/java/io/pravega/shared/segment/StreamSegmentNameUtils.java [hunk 4]
```diff
@@ -0,0 +1,486 @@
+/**
```
Comment (sachin-j-joshi) [inline shared/protocol/src/main/java/io/pravega/shared/segment/StreamSegmentNameUtils.java]: Not sure why it is showing as new file. This probably means you'll have  to deal with a merge conflict.

File: bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3Storage.java [hunk 5]
```diff
@@ -13,12 +13,7 @@
 import com.emc.object.s3.S3Client;
 import com.emc.object.s3.S3Exception;
 import com.emc.object.s3.S3ObjectMetadata;
-import com.emc.object.s3.bean.AccessControlList;
-import com.emc.object.s3.bean.CanonicalUser;
-import com.emc.object.s3.bean.CopyPartResult;
-import com.emc.object.s3.bean.Grant;
-import com.emc.object.s3.bean.MultipartPartETag;
-import com.emc.object.s3.bean.Permission;
+import com.emc.object.s3.bean.*;
```
Comment (sachin-j-joshi) [inline bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3Storage.java:16]: The checkStyle rule will fail on this commit.
Style rule prohibits using * imports. 
I suggest you run `./gradlew compileJava compileTest checkStyleMain checkStyleTest spotBugsMain spotbugsTest` to find all issues and fix.  
(In IdeaJ when you hover over * import the "lighbulb" hint shows option to replace with single imports.)
Please fix the issues.

**Comment by naveenkothamasu** [comment]
Moved to https://github.com/pravega/pravega/pull/4678

**Pull request closed without merging**